[JSC] Support all rounding modes in Temporal

rkirsling opened the pull request:
#### a8f81aab31317cd32026b3f895f309e02bc9089f
<pre>
[JSC] Support all rounding modes in Temporal
<a href="https://bugs.webkit.org/show_bug.cgi?id=245935">https://bugs.webkit.org/show_bug.cgi?id=245935</a>

Reviewed by Geoffrey Garen.

This patch implements the rest of the rounding modes for Temporal, namely,
`expand`, `halfCeil`, `halfFloor`, `halfTrunc`, and `halfEven`.

(As a bonus, it also addresses two adjacent spec tweaks related to fractionalSecondDigits and BalanceDuration.)

Note: Some tests still fail due to precision loss beyond MAX_SAFE_INTEGER...

* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::TemporalDuration::balance):
(JSC::TemporalDuration::total const):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::temporalFractionalSecondDigits):
(JSC::temporalRoundingMode):
(JSC::negateTemporalRoundingMode):
(JSC::roundNumberToIncrement):

Canonical link: <a href="https://commits.webkit.org/255068@main">https://commits.webkit.org/255068@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37bf979bb99dd2e9ffa2e484db17e48771210f56

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91193 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/193 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21769 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100798 "Hash 37bf979b for PR 4907 does not build (failure)") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/160514 "Reverted pull request changes (failure)") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/205 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29216 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83555 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97331 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96849 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/167 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/77936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27139 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82112 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70169 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/82665 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35339 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15787 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/77689 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33135 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16824 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26792 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3519 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36919 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/77936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80289 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38846 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35915 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17600 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/JavaScriptCore/runtime/TemporalDuration.cpp:
```diff
@@ -549,7 +553,9 @@ double TemporalDuration::total(JSGlobalObject* globalObject, JSValue optionsValu
     ISO8601::Duration newDuration = m_duration;
-    balance(newDuration, unit);
+    auto infiniteResult = balance(newDuration, unit);
+    if (infiniteResult)
```
I kind of prefer "infiniteResult.has_value()" in code like this. Otherwise it's too easy to read this as a check for zero. I'm not sure this is a project-wide consensus though.

rkirsling replied:
Hmm, seems like this could be one of those JSC vs. everybody else things. 😅 `.has_value())` (_with trailing paren_) has nearly no results in JSC, despite having a bunch in WebCore & WebKit. Since we've been using this style quite a bit in neighboring code, I'm inclined to leave it as-is, as long as you don't object.

geoffreygaren reviewed (approved):
r=me

rkirsling committed:
[JSC] Support all rounding modes in Temporal
https://bugs.webkit.org/show_bug.cgi?id=245935

Reviewed by Geoffrey Garen.

This patch implements the rest of the rounding modes for Temporal, namely,
`expand`, `halfCeil`, `halfFloor`, `halfTrunc`, and `halfEven`.

(As a bonus, it also addresses two adjacent spec tweaks related to fractionalSecondDigits and BalanceDuration.)

Note: Some tests still fail due to precision loss beyond MAX_SAFE_INTEGER...

* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::TemporalDuration::balance):
(JSC::TemporalDuration::total const):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::temporalFractionalSecondDigits):
(JSC::temporalRoundingMode):
(JSC::negateTemporalRoundingMode):
(JSC::roundNumberToIncrement):

Canonical link: https://commits.webkit.org/255068@main
